File: kg/kgClient.ts
export class KGClient {
  constructor(private base = 'http://localhost:4545') {
    this.base = this.base.replace(/\/$/, '');
  }

  async upsert(type: string, props: any) {
    await fetch(`${this.base}/entity`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ type, props }),
    });
  }

  async link(srcId: string, predicate: string, dstId: string) {
    await fetch(`${this.base}/link`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ srcId, predicate, dstId }),
    });
  }

  async find(query: Record<string, unknown>) {
    const res = await fetch(`${this.base}/find`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(query),
    });
    if (!res.ok) {
      throw new Error(`find failed: ${res.status}`);
    }
    return res.json();
  }

  async lineage(id: string, direction: 'up' | 'down' = 'up', maxDepth = 3) {
    const res = await fetch(`${this.base}/lineage/${encodeURIComponent(id)}?direction=${direction}&maxDepth=${maxDepth}`);
    if (!res.ok) {
      throw new Error(`lineage failed: ${res.status}`);
    }
    return res.json();
  }
}

File: kg/kg_client.py
import requests
import urllib.parse as up


class KG:
    def __init__(self, base: str = "http://localhost:4545") -> None:
        self.base = base.rstrip('/')

    def upsert(self, type_, props):
        r = requests.post(f"{self.base}/entity", json={"type": type_, "props": props}, timeout=10)
        r.raise_for_status()
        return r.json() if r.text else {"ok": True}

    def link(self, src, pred, dst):
        payload = {"srcId": src, "predicate": pred, "dstId": dst}
        r = requests.post(f"{self.base}/link", json=payload, timeout=10)
        r.raise_for_status()
        return r.json() if r.text else {"ok": True}

    def find(self, **query):
        r = requests.post(f"{self.base}/find", json=query, timeout=10)
        r.raise_for_status()
        return r.json()

    def lineage(self, id_, direction="up", max_depth=3):
        url = f"{self.base}/lineage/{up.quote(id_)}?direction={direction}&maxDepth={max_depth}"
        r = requests.get(url, timeout=10)
        r.raise_for_status()
        return r.json()
